### ISSUE TITLE
Native memory overuse by DirectBuffers when using the ember server

### ISSUE THREAD

**Issue by Bunyod** [http4s/http4s#7343]
Recently while updating our services we bumped our dependencies to the latest versions of Typelevel Stack and replaced the http4s backend Blaze with Ember as suggested in the official documentation.

After the upgrade, we started receiving alerts that our services were being restarted from time to time.

We took a close look into metrics and noticed that the services were killed due to native memory usage: resident set size (RSS) reached the native memory limit, while heap memory was not close to the limit (see figure below). OOM exception was never recorded.
![grafana_dashboard](https://github.com/http4s/http4s/assets/5593355/70144553-9a27-4b65-907a-ee5c76b8efe8)
We analyzed the memory and thread dumps of the services and didn’t find any clear indication of a memory leak.

Based on our observations we noticed an odd behavior of the direct buffers, so we decided to compare the http4s backends blaze vs ember.

For each http4s backend, we defined a dummy http4s server with one simple endpoint (/ping → OK(”pong”) and we simulated a huge number of concurrent requests (1M total requests, thread pool size of 10). While doing so we recorded the Direct Buffer behavior via VisualVM (see figures below).
![blazemainjmx](https://github.com/http4s/http4s/assets/5593355/fd6a0bbf-8759-4235-8c1c-a0a6ae7f66e3)
![EmberMainJMX](https://github.com/http4s/http4s/assets/5593355/5b90c46e-df95-4368-8e81-8e285c6d5c3b)
<aside>
💡 As can be noticed blaze keeps a stable behavior while ember keeps increasing the DirectBuffer. 
If the DirectBuffers are not released by a Garbage Collector cycle fast enough (e.g. there is still enough heap memory available) this leads to process termination, as described above.

</aside>

Here is a GitHub repo to reproduce our experiments: 
https://github.com/Bunyod/ember-fs2-demo

This is a possible place to dig in deeper about the cause of the problem:

https://github.com/typelevel/fs2/blob/v3.9.3/io/jvm-native/src/main/scala/fs2/io/net/SocketPlatform.scala#L46C26-L49C1

**Versions**:

java version: openjdk-11.0.14-jdk-slim

Ember dummy project versions:
cats-effect:3.5.2
http4s-ember-server:0.23.24
fs2:3.9.3

Blaze dummy project versions:
http4s-blaze-server:0.23.15

**Comment by armanbilge** [comment]
Thanks for the report!

I see that you are using `HttpURLConnection` in your reproducer. Out of curiosity, can you still replicate the issue if you use the [`HttpClient`](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpClient.html)? Thanks!

**Comment by Bunyod** [comment]
I've tried to use http4s-ember-client, but the performance gets even worse
![tenkreq](https://github.com/http4s/http4s/assets/5593355/370566b2-3218-4f1d-a4fb-a8e8a5278d9f)

Maybe I'm misusing it, [here is the example code](https://github.com/Bunyod/ember-fs2-demo/blob/main/ember-demo/src/main/scala/com/ember/fs2/demo/EmberTestEndpointIO.scala).

**Comment by anesto-deltoro** [comment]
It is worth stressing that if we force a GC run the memory is released which explains why we couldn't find any hints of a memory leak during our analysis of the thread dumps. The restarts occur when the JVM heap is not under pressure, during which there is no GC cycle triggered, but the native memory reaches the limit due to the increasing growth of DirectBuffers.

**Comment by danicheg** [comment]
It'd be helpful to share your JVM parameters (memory, GC, maybe something else).

**Comment by anesto-deltoro** [comment]
@danicheg here are some details of the configuration we use

**Kubernetes Pod configuration:**
Limits:
	memory:  2Gi
Requests:
	memory:   2Gi
	cpu:      500m

**JAVA_OPTS:**
-XX:+PrintFlagsFinal
-Djava.rmi.server.hostname=127.0.0.1
-server
-Xms1340M
-Xmx1340M
-XX:MaxMetaspaceSize=256M
-XX:MetaspaceSize=256M
-XX:ReservedCodeCacheSize=240M
-Xss1M
-XX:MaxGCPauseMillis=100
-XX:+UseG1GC

**Comment by armanbilge** [comment]
> is there any configuration on ember server side to limit the number of opened sockets?

https://github.com/http4s/http4s/blob/4cf0a9f8e6cb585493469b4a96a924697cf7fe1a/ember-server/shared/src/main/scala/org/http4s/ember/server/EmberServerBuilder.scala#L164-L165

But even if you put a low limit on this, you might still be able to replicate the issue by rapidly opening/closing a single connection over and over.

**Comment by anesto-deltoro** [comment]
Follow up after holidays...
@armanbilge We have no control over how our APIs are used by external clients, so we can't rely on/assume optimal use of connection pools.

We took a look into the [PR](https://github.com/typelevel/fs2/pull/2212) that introduced the use of direct buffers, and as noted in the comments by djspiewak the performance gains were marginal (almost no difference).

We created an internal fork of fs2 and replaced the use of direct buffers with indirect buffers and we confirmed that the memory is stable/predictable and our services are not restarting anymore.
The aforementioned change is limited to replacing 3 occurrences of ByteBuffer.allocateDirect with ByteBuffer.allocate.

At this point, we would like to know how to proceed. It was mentioned that a benchmark to determine whether there is a performance benefit is needed. We understand that the usage of direct buffers should offer a performance benefit, but the unpredictable behavior of memory use is probably a high price to pay, especially if it is traduced into service restarts on production.
